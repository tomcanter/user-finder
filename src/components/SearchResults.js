// @flow

import React, {
  PropTypes,
} from 'react';
import curry from 'lodash/fp/curry';
import map from 'lodash/fp/map';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';
import Result from './Result';

type Props = {
  entities: Object,
  searchTerm: string,
  ids: Array<number>,
  total: number,
};

const renderSearchResult = curry((entities, item) => {
  const {
    id,
    avatar_url: avatarUrl,
    login,
  } = entities[item];
  return (
    <li key={id}>
      <Result
        username={login}
        avatarUrl={avatarUrl}
      />
    </li>
  );
});

function renderResultsMessage(searchTerm: string, total: number, resultLength: number) {
  if (!searchTerm || !resultLength) {return null;}
  return (
    <p><b>{searchTerm}</b> found {total} {total === 1 ? 'result' : 'results'}</p>
  );
}

function SearchResults({entities, ids, searchTerm, total}: Props) {
  return (
    <div>
      {renderResultsMessage(searchTerm, total, ids.length)}
      <ul className={css(styles.SearchResults_List)}>
        {map(renderSearchResult(entities), ids)}
      </ul>
    </div>
  );
}

SearchResults.propTypes = {
  entities: PropTypes.object,
  searchTerm: PropTypes.string,
  ids: PropTypes.array,
  total: PropTypes.number,
};

const styles = StyleSheet.create({
  SearchResults_List: {
    display: 'flex',
    flexWrap: 'wrap',
  },
});

export default SearchResults;
