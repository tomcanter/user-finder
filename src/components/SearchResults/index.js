// @flow

import React, {
  PropTypes,
} from 'react';
import curry from 'lodash/fp/curry';
import map from 'lodash/fp/map';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';
import Result from 'components/Result';

type Props = {
  entities: Object,
  searchTerm: string,
  ids: Array<number>,
  total: number,
};

const renderSearchResult = curry((entities, item) => {
  const {
    id,
    avatar_url: avatarUrl,
    login,
  } = entities[item];
  return (
    <li key={id} className={css(styles.SearchResults_item)}>
      <Result
        username={login}
        avatarUrl={avatarUrl}
      />
    </li>
  );
});

function renderResultsMessage(searchTerm: string, total: number, resultLength: number) {
  if (!searchTerm || !resultLength) {return null;}
  return (
    <p><b>{searchTerm}</b> found {total} {total === 1 ? 'result' : 'results'}</p>
  );
}

function SearchResults({entities, ids, searchTerm, total}: Props) {
  return (
    <div className={css(styles.SearchResults)}>
      {renderResultsMessage(searchTerm, total, ids.length)}
      <ul className={css(styles.SearchResults_list)}>
        {map(renderSearchResult(entities), ids)}
      </ul>
    </div>
  );
}

SearchResults.propTypes = {
  entities: PropTypes.object,
  searchTerm: PropTypes.string,
  ids: PropTypes.array,
  total: PropTypes.number,
};

const BREAKPOINT = '@media (min-width: 600px)';

const styles = StyleSheet.create({
  SearchResults: {
    maxWidth: 1280,
    margin: '0 auto',
  },

  SearchResults_list: {
    display: 'flex',
    flexWrap: 'wrap',
  },

  SearchResults_item: {
    flex: '1 0 auto',

    [BREAKPOINT]: {
      ':only-child': {
        flex: 0,
      },
    },
  },
});

export default SearchResults;
