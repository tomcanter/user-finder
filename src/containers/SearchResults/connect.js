// @flow

import {connect} from 'react-redux';
import get from 'lodash/fp/get';

export function mapStateToProps(state: Object): Object {
  return {
    userEntities: get('entities.users', state),
    userIds: get('search.result', state),
    totalResults: get('search.totalResults', state),
    pagination: get('search.pagination', state),
  };
}

export default connect(mapStateToProps);
