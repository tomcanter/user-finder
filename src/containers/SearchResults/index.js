// @flow

import React from 'react';
import {
  StyleSheet,
} from 'aphrodite/no-important';
import SearchResults from 'components/SearchResults';
import Pagination from 'components/Pagination';
import ResultsMessage from 'components/ResultsMessage';
import Loading from 'components/Loading';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  isPending: boolean,
  pagination: Object,
  searchTerm: string,
  totalResults: number,
  userEntities: Object,
  userIds: Array<number>,
};

export function SearchResultsContainer(props: Props) {
  const {
    isPending,
    pagination,
    searchTerm,
    totalResults,
    userEntities,
    userIds,
  } = props;

  if (!searchTerm) {return null;}

  if (isPending) {
    return <Loading />;
  }

  return (
    <div>
      <Container
        rootStyle={styles.SearchResultsContainer_message}
      >
        <ResultsMessage
          searchTerm={searchTerm}
          resultsTotal={totalResults}
          pageTotal={userIds.length}
        />
      </Container>
      <Container
        noGutter={true}
        rootStyle={styles.SearchResultsContainer_results}
      >
        <SearchResults
          searchTerm={searchTerm}
          ids={userIds}
          entities={userEntities}
          total={totalResults}
        />
      </Container>
      <Container>
        <Pagination {...pagination} />
      </Container>
    </div>
  );
}

SearchResultsContainer.defaultProps = {
  pagination: {},
  searchTerm: '',
};

const styles = StyleSheet.create({
  SearchResultsContainer_results: {
    marginTop: 15,
  },

  SearchResultsContainer_message: {
    marginTop: 15,
  },
});

export default connect(SearchResultsContainer);
