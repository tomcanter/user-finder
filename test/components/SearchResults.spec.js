import React from 'react';
import renderer from 'react-test-renderer';

import SearchResults from '../../src/components/SearchResults';

describe('Component: SearchResults', () => {

  describe('when given a set of results and entities', () => {
    it('should render them with a message', () => {
      const props = {
        entities: {
          1: {login: 'foo', avatar_url: 'avatar.jpg', id: 1},
          2: {login: 'foobar', avatar_url: 'avatar.jpg', id: 2},
        },
        ids: [1, 2],
        searchTerm: 'foo',
        total: 2,
      };
      const component = renderer.create(
        <SearchResults {...props} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('when given a one result', () => {
    it('should render a correctly worded message', () => {
      const props = {
        entities: {
          1: {login: 'foo', avatar_url: 'avatar.jpg', id: 1},
        },
        ids: [1],
        searchTerm: 'foo',
        total: 1,
      };
      const component = renderer.create(
        <SearchResults {...props} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('when the results are 0', () => {
    it('should not render a message', () => {
      const props = {
        entities: {},
        ids: [],
        searchTerm: 'foo',
        total: 0,
      };
      const component = renderer.create(
        <SearchResults {...props} />
      );
      expect(component).toMatchSnapshot();
    });
  });

});
