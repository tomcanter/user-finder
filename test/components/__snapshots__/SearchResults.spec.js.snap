exports[`Component: SearchResults when given a one result should render a correctly worded message 1`] = `
<div>
  <p>
    <b>
      foo
    </b>
     found 
    1
     
    result
  </p>
  <ul
    className="SearchResultsList_96vuga">
    <li>
      <div>
        <p>
          foo
        </p>
        <img
          alt="foo"
          height={130}
          src="avatar.jpg"
          width={130} />
      </div>
    </li>
  </ul>
</div>
`;

exports[`Component: SearchResults when given a set of results and entities should render them with a message 1`] = `
<div>
  <p>
    <b>
      foo
    </b>
     found 
    2
     
    results
  </p>
  <ul
    className="SearchResultsList_96vuga">
    <li>
      <div>
        <p>
          foo
        </p>
        <img
          alt="foo"
          height={130}
          src="avatar.jpg"
          width={130} />
      </div>
    </li>
    <li>
      <div>
        <p>
          foobar
        </p>
        <img
          alt="foobar"
          height={130}
          src="avatar.jpg"
          width={130} />
      </div>
    </li>
  </ul>
</div>
`;

exports[`Component: SearchResults when the results are 0 should not render a message 1`] = `
<div>
  <ul
    className="SearchResultsList_96vuga" />
</div>
`;
